### PR TITLE
square support

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77499
* #77740
* #77739
* #77681
* __->__ #77682

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -378,7 +378,7 @@ def _isnan(a: Tensor) -> Tensor:
 square = _make_elementwise_unary_reference(
     prims.square,
     type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.BOOL_TO_LONG,
-    aten_op=None,  # CompositeImplicitAutograd
+    aten_op=torch.ops.aten.square.out,
```
Comment (ngimel) [inline torch/_refs/__init__.py:381]: why do we need `out` overload here?

Comment (ezyang) [reply torch/_refs/__init__.py:381]: It's composite implicit autograd but I guess we didn't do the out variant correctly, not sure

**Review by ngimel — APPROVED** [review]

File: torch/_refs/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/_refs/__init__.py:381]:
```
- func: square(Tensor self) -> Tensor
  device_check: NoCheck   # TensorIterator
  variants: function, method

- func: square_(Tensor(a!) self) -> Tensor(a!)
  device_check: NoCheck   # TensorIterator
  variants: function, method

- func: square.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
  dispatch:
    CPU, CUDA: square_out
    MPS: square_out_mps
    ```

TBH this is kind of weird.

Comment (ngimel) [reply torch/_refs/__init__.py:382]: Yeah, CPU, CUDA dispatch for `out` is wrong, it calls `return at::pow_out(result, self, 2);`, should still be composite implicit autograd.

**Review by mruberry — APPROVED** [review]
Stamped!

**Pull request closed without merging**